### DISCUSSION TITLE
how to intertwine HTTP call to API with my Drift isolate

### DISCUSSION THREAD

**Discussion by matthewkooshad** [simolus3/drift#2491] [Q&A] [answered]
I've read over https://drift.simonbinder.eu/docs/advanced-features/isolates/ .
I'm noticing that when I make and await on a http request, it seems some things are not waiting. And, I think that's probably because I am using an isolate.
I really like the idea that using an isolate, I don't block the main thread.
But, my goal is to share data between sqlite and the data behind my web api, having them in sync. 
any advice on how to handle awaiting on the http call, so that I can collaborate the http call with the DB transaction? 
it doesn't seem to be playing well together with what I am trying, simply using await, and using an isolate. 
thanks!

**Comment by matthewkooshad** [comment]
maybe I should run the Drift part, then use asyncMap to run the http part to get the real ID for the entry from the web api? the idea I am playing with here also is that the Drift part could execute fine, but if the device is offline, the http web api part needs to wait until later.

**Comment by simolus3** [comment, marked as answer]
Could you share more details about how you're making the HTTP call at the moment and which drift operation it seems to be blocking?

As the isolate documentation points out, most apps don't need to manage drift isolates manually. Instead, use a `NativeDatabase.createInBackground` when constructing your database class to let drift manage a isolate for you.

Running HTTP calls inside a transaction is not ideal since it will block everything else on the database, regardless of whether isolates are used or not. So don't do something like

```dart
await transaction(() async {
  final rows = await fetchRowsFromServer();
  await insertRows(rows);
});
```

Instead, move the `fetchRowsFromServer()` call outside of the transaction and only run the database operation itself in the transaction.

-------

Note that HTTP calls are mostly asynchronous too, moving them to a background isolate is only useful when there's a lot of JSON parsing. In that case, you can use the [simple sharing](https://drift.simonbinder.eu/docs/advanced-features/isolates/#simple-sharing) feature to create a short-lived isolate to make the HTTP request and to insert rows into the database. The advice about not doing that in a `transaction` callback still applies though.

**Reply by matthewkooshad** [reply]
thanks Simon for your answer with the details. in facing the issue, i really just refactored the approach, and decided to attempt the http separately after the sqlite transaction is finished. i wish i had some detail to share here, but i kept working at it, trying to deal with it, so i didn't have a decent sample to share on what was giving me a roadblock. thanks for always being so helpful!
